### PR TITLE
Overhaul Unicode preferences

### PR DESCRIPTION
**Pull request by lgoettgens**
To be able to disable unicode for tests/doctests etc., see e.g. https://github.com/oscar-system/Oscar.jl/pull/3271

Resolves https://github.com/Nemocas/AbstractAlgebra.jl/issues/1588.

### CODE REVIEW

**Comment by thofma** [comment]
No clue (neither what the problem nor what the proposed solution is).

Does that address your concerns @benlorenz?

File: src/PrettyPrinting.jl [hunk 1]
```diff
@@ -1386,24 +1386,66 @@ end
...
-# A non-compiletime preference
-function allow_unicode(flag::Bool)
-  old_flag = is_unicode_allowed()
-  @set_preferences!("unicode" => flag)
-  return old_flag
+@static if VERSION < v"1.8.0-DEV.1465"
+   ALLOW_UNICODE_OVERRIDE_VALUE = nothing
+else
+   ALLOW_UNICODE_OVERRIDE_VALUE::Union{Bool,Nothing} = nothing
 end
```
Comment (fingolfin) [inline src/PrettyPrinting.jl:1393]: Could also do
```suggestion
const ALLOW_UNICODE_OVERRIDE_VALUE = Ref{Union{Bool,Nothing}}(nothing)
```
and then access it via `ALLOW_UNICODE_OVERRIDE_VALUE[]`

File: src/PrettyPrinting.jl [hunk 2]
```diff
@@ -1386,24 +1386,66 @@ end
...
+@doc """
+    allow_unicode(flag::Bool; temporary::Bool=false) -> Bool
+
+Set whether unicode characters are allowed in pretty printing and returns the
+previous value.
+If `temporary` is `true`, then the change is only active for the current worker and session.
+Otherwise, the change is permanently saved in the preferences.
+A permanent change will always override a temporary change.
+
+This function may behave arbitrarily, if called from within an argument to
```
Comment (fingolfin) [inline src/PrettyPrinting.jl:1404]:
```suggestion
This function may behave arbitrarily if called from within an argument to
```

File: src/PrettyPrinting.jl [hunk 3]
```diff
@@ -1386,24 +1386,66 @@ end
...
 
-function with_unicode(f::Function)
-  old_allow_unicode = allow_unicode(true)
-  try
-    f()
-  finally
-    allow_unicode(old_allow_unicode)
-  end
+@doc """
+    with_unicode(f::Function, flag::Bool=true)
```
Comment (fingolfin) [inline src/PrettyPrinting.jl:1434]: Maybe rename `flag` to `allowed`, then it is becomes quite clear what it does?

File: src/PrettyPrinting.jl [hunk 4]
```diff
@@ -1386,24 +1386,66 @@ end
...
+
+Temporarily set whether unicode characters are allowed in pretty printing
+during the execution of `f`.
+"""
+function with_unicode(f::Function, flag::Bool=true)
+   global ALLOW_UNICODE_OVERRIDE_VALUE
+   previous = ALLOW_UNICODE_OVERRIDE_VALUE
+   ALLOW_UNICODE_OVERRIDE_VALUE = flag
+   try
+      f()
```
Comment (fingolfin) [inline src/PrettyPrinting.jl:1444]: Seems we have no tests for `with_unicode` in AA` :-(. But of course this is out of scope for this PR.

**Review by fingolfin — APPROVED** [review]
Some optional suggestions added

File: src/PrettyPrinting.jl (hunk 2 above)
Comment (fingolfin) [reply src/PrettyPrinting.jl:1404]: Actually thinking about this some more, I find "within an argument" a bit confusing. Maybe "within scope of a `with_unicode` do-block" -- that one can use other function beyond do-blocks is clear to the experts? Or we write "within scope of a function executed by `with_unicode`" ?

File: src/PrettyPrinting.jl (hunk 1 above)
Comment (lgoettgens) [reply src/PrettyPrinting.jl:1393]: `Ref`s are black magic for me, so I would rather keep it, and simplify once we drop 1.6 supprot.

File: src/PrettyPrinting.jl (hunk 2 above)
Comment (lgoettgens) [reply src/PrettyPrinting.jl]: reformulated

File: src/PrettyPrinting.jl (hunk 4 above)
Comment (lgoettgens) [reply src/PrettyPrinting.jl:1452]: That`s true. But after thinking again about it, this seems to be kind of hard anyway:
Calling `allow_unicode()` from the test would change the user preferences while running tests (not a great experience for devs), and from a within `with_unicode` block `allow_unicode` is unsafe.

But overall, we should postpone it

File: src/PrettyPrinting.jl (hunk 1 above)
Comment (fingolfin) [reply src/PrettyPrinting.jl:1393]: A Ref is basically a vector of fixed length 1 :-)

**Comment by lgoettgens** [comment]
All of this is non-breaking and can thus be merged, right?

File: src/PrettyPrinting.jl [hunk 5]
```diff
@@ -1386,24 +1386,74 @@ end
...
+```julia
+with_unicode([allowed]) do
+   # code that should be executed with unicode allowed/disallowed
+end
+```
+"""
+function with_unicode(f::Function, allowed::Bool=true)
+   global ALLOW_UNICODE_OVERRIDE_VALUE
+   previous = ALLOW_UNICODE_OVERRIDE_VALUE
+   ALLOW_UNICODE_OVERRIDE_VALUE = allowed
```
[the remark this answers is not in the document]
Comment (lgoettgens) [reply src/PrettyPrinting.jl:1450]: no, that does not work here. imagine that the override value is not set, but only the preference file. This current pattern correctly sets the override value to `nothing` afterward, but your suggestion sets it to the value of the preference file.
